lib/include/jxl/decode.h

```diff
@@ -746,6 +746,56 @@ JXL_EXPORT JxlDecoderStatus JxlDecoderGetColorAsICCProfile(
 JXL_EXPORT JxlDecoderStatus JxlDecoderSetPreferredColorProfile(
     JxlDecoder* dec, const JxlColorEncoding* color_encoding);
 
+/**
+ * Sets the color management system (CMS) that will be used for color conversion
+ * (if applicable) during decoding. May only be set before starting decoding. If
+ * left unset, the decoder will not do arbitrary color conversions:
+ * JxlDecoderSetPreferredColorProfile can be used, but there is no guarantee
+ * that the decoded image will be returned in the preferred color encoding. If
+ * however a CMS is set using this function, then the function
+ * JxlDecoderSetColorEncoding or JxlDecoderSetICCProfile can be used and the
+ * decoded image will then be guaranteed to be in the requested color encoding.
+ *
+ * @param dec decoder object.
+ * @param cms structure representing a CMS implementation. See JxlCmsInterface
+ * for more details.
+ */
+JXL_EXPORT void JxlDecoderSetCms(JxlDecoder* dec, JxlCmsInterface cms);
+
+/**
+ * Sets the desired output color encoding of the decoded image. This function
+ * can only be used if JxlDecoderSetCms was previously set. This is an
+ * alternative to JxlDecoderSetICCProfile and only one of these two must be
+ * used. This one sets the color encoding as a @ref JxlColorEncoding, while the
+ * other sets it as ICC binary data.
+ *
+ * @param dec decoder object.
+ * @param color color encoding. Object owned by the caller and its contents are
+ * copied internally.
+ * @return JXL_DEC_SUCCESS if the operation was successful, JXL_DEC_ERROR
+ * otherwise.
+ */
+JXL_EXPORT JxlDecoderStatus
+JxlDecoderSetColorEncoding(JxlDecoder* dec, const JxlColorEncoding* color);
+
+/**
+ * Sets the desired output color encoding of the decoded image, as an ICC color
+ * profile. This function can only be used if JxlDecoderSetCms was previously
+ * set. The ICC profile has to be a valid RGB color profile (not grayscale or
```

Comment:
What is the problem with grayscale profiles so they cannot be allowed here?

Reply:
It's not strictly a problem but I think it will be convenient if only the code paths to RGB have to be implemented: gray to RGB, RGB to RGB, CMYK to RGB. If we allow gray output encodings, it doubles the number of cases...

I think in most general-purpose applications you want to be able to decode both color and grayscale images, and the target display space is RGB, so I think you'll want to set this to whatever the display space is (probably sRGB, Display P3, or Rec2100 PQ in most cases).

For special-purpose applications, I think they can do the appropriate color management themselves.

Reply:
I think it would make sense to allow Gray -> Gray at least, so you could get the pixel data in the TARGET_ORIGINAL space in the case of grayscale.

Reply:
Agreed, Gray to Gray does make sense. So let's say the following are allowed:

- RGB to RGB
- CMYK to RGB
- Gray to RGB
- Gray to Gray

I don't think we want to allow conversions _to_ CMYK; perhaps it is also useful to allow conversions from RGB/CMYK to Gray but I think that's probably not a big use case and I want to avoid having to implement more code paths than needed.

In case the original space is a CMYK space, I propose we only allow you to get the pixels converted to an RGB space, or else to get them as they are encoded (which is the original space in the non-XYB case). Probably we should not allow the combination of XYB and CMYK — I think the spec currently allows that, but it's ambiguous how to interpret such data.

Reply:
What do we do now here, @jonsneyers?

Reply:
I think that what I wrote in the comment above back in May still makes sense. At least I think that's the minimum we should allow/implement. If it's not much trouble to just allow anything the cms can handle (including stuff like decoding an RGB image to CMYK, if the cms can do that), then that's also fine for me.

For any viewing application, they need to be able to convert any jxl to the display colorspace, or at least to an RGB(A) buffer plus an ICC profile that they can then pass to something that knows how to deal with that. So (anything) to RGB is needed.

For grayscale images, it can be annoying to force them to be converted to RGB since that requires a buffer 3 times as large. So gray to gray is useful too imo.

Reply:
Explained it like that in the doc string now.